src/Process/Runner/MutationTestingRunner.php

```diff
@@ -58,45 +60,82 @@ final class MutationTestingRunner
+        EventDispatcher $eventDispatcher,
+        bool $runConcurrently
     ) {
         $this->processBuilder = $mutantProcessBuilder;
         $this->mutantFactory = $mutantFactory;
         $this->parallelProcessManager = $parallelProcessManager;
         $this->eventDispatcher = $eventDispatcher;
+        $this->runConcurrently = $runConcurrently;
     }
 
     /**
      * @param Mutation[] $mutations
      */
-    public function run(array $mutations, int $threadCount, string $testFrameworkExtraOptions): void
+    public function run(iterable $mutations, int $threadCount, string $testFrameworkExtraOptions): void
     {
-        $this->eventDispatcher->dispatch(new MutantsCreationWasStarted(count($mutations)));
+        $numberOfMutants = $this->bufferAndCountIfNeeded($mutations);
 
-        $processes = array_map(
-            function (Mutation $mutation) use ($testFrameworkExtraOptions): MutantProcess {
-                $mutant = $this->mutantFactory->create($mutation);
+        $this->eventDispatcher->dispatch(new MutantsCreationWasStarted($numberOfMutants));
 
-                $process = $this->processBuilder->createProcessForMutant($mutant, $testFrameworkExtraOptions);
+        $processes = take($mutations);
+        $processes->map(function (Mutation $mutation) use ($testFrameworkExtraOptions): MutantProcess {
+            $mutant = $this->mutantFactory->create($mutation);
 
-                $this->eventDispatcher->dispatch(new MutantWasCreated());
+            $process = $this->processBuilder->createProcessForMutant($mutant, $testFrameworkExtraOptions);
 
-                return $process;
-            },
-            $mutations
-        );
+            $this->eventDispatcher->dispatch(new MutantWasCreated());
+
+            return $process;
+        });
 
         $this->eventDispatcher->dispatch(new MutantsCreationWasFinished());
+        $this->eventDispatcher->dispatch(new MutationTestingWasStarted($numberOfMutants));
+
+        // We filter these here because it is beyond responsibilities of a process manager to care about types of processes
+        $processes = take($processes)->filter(function (MutantProcess $mutantProcess) {
+            if ($mutantProcess->getMutant()->isCoveredByTest()) {
+                return true;
+            }
+
+            $this->eventDispatcher->dispatch(new MutantProcessWasFinished(
+                MutantExecutionResult::createFromProcess($mutantProcess)
+            ));
 
-        $this->eventDispatcher->dispatch(new MutationTestingWasStarted(count($processes)));
+            return false;
+        });
 
         $this->parallelProcessManager->run($processes, $threadCount);
 
         $this->eventDispatcher->dispatch(new MutationTestingWasFinished());
     }
+
+    /**
+     * @param iterable|mixed[] $subjects
+     */
+    private function bufferAndCountIfNeeded(iterable &$subjects): int
+    {
+        if ($this->runConcurrently) {
+            return 0;
```

Comment:
I'm not sure to understand: why `0` here?

Reply:
It'll be fed to `MutationTestingWasStarted` and later fed to an `OutputFormatter` like `DotFormatter` which wants an `int`.   This isn't the cleanest solution, sure, but that's avoids changing the event system even further.

Reply:
Would `-1` do the job as well? Because `0` is a legit values so `-1` would have the merit to be a bit more conventional to say "there's no count"

Reply:
If it's not I'll change that in a dedicated PR after this one is merge, no biggie